exposure-notification/src/main/java/fi/thl/covid19/exposurenotification/diagnosiskey/DiagnosisKeyDao.java

```diff
@@ -118,6 +125,81 @@ public boolean verify(int verificationId, String requestChecksum) {
+        }
+    }
+
+    public List<TemporaryExposureKey> fetchAvailableKeysForEfgs(long operationId) {
+        LOG.info("Fetching keys not sent to efgs.");
+        String sql = "select key_data, rolling_period, rolling_start_interval_number, transmission_risk_level, submission_interval " +
+                "from en.diagnosis_key " +
+                "where efgs_operation = :efgs_operation " +
+                "order by key_data";
+        return jdbcTemplate.queryForList(sql, Map.of("efgs_operation", operationId), TemporaryExposureKey.class);
+    }
+
+    public void finishOperation(long operationId) {
+        String sql = "update en.efgs_operation set state = :new_state where id = :id";
+        jdbcTemplate.update(sql, Map.of("new_state", EfgsOperationState.FINISHED, "id", operationId));
+    }
+
+    public void markErrorOperation(long operationId) {
+        String sql = "update en.efgs_operation set state = :error_state where id = :id";
+        jdbcTemplate.update(sql, Map.of("error_state", EfgsOperationState.ERROR, "id", operationId));
+    }
+
+    @Transactional
+    public void addInboundKeys(List<TemporaryExposureKey> keys, int interval) {
+        if (!keys.isEmpty()) {
+            long operationId = startInboundOperation();
+            try {
+                batchInsert(interval, keys, operationId);
+                LOG.info("Inserted keys: {} {}", keyValue("interval", interval), keyValue("count", keys.size()));
+                finishOperation(operationId);
+            } catch (Throwable t) {
+                markErrorOperation(operationId);
+                throw new EfgsOperationException("Inbound operation from efgs failed.", t);
+            }
+        }
+    }
+
+    private long startInboundOperation() {
+        KeyHolder operationKeyHolder = new GeneratedKeyHolder();
+
+        String createOperation = "insert into en.efgs_operation (direction) values (:direction)";
+        jdbcTemplate.update(createOperation, new MapSqlParameterSource("direction", EfgsOperationDirection.INBOUND), operationKeyHolder);
+        return requireNonNull(operationKeyHolder.getKey()).longValue();
+    }
+
+    private boolean isOutboundOperationAvailable() {
+        String sql = "select count(*) from en.efgs_operation where state = :state";
+        return jdbcTemplate.query(sql, Map.of("state", EfgsOperationState.STARTED), (rs, i) -> rs.getInt(1))
+                .stream().findFirst().orElseThrow(() -> new IllegalStateException("Count returned nothing.")) == 0;
+    }
+
+    private long markOutboundOperationStarted() {
+        KeyHolder operationKeyHolder = new GeneratedKeyHolder();
+
+        String createOperation = "insert into en.efgs_operation (direction) values (:direction)";
+        jdbcTemplate.update(createOperation, new MapSqlParameterSource("direction", EfgsOperationDirection.OUTBOUND), operationKeyHolder);
+
+        String markToBeUpdated = "update en.diagnosis_key " +
```

Comment:
We have inbound and outbound keys, so we need to handle the integration state by key in some manner. Also, we can add some metadata in operation table. There is different angles in this, performance and maintainability. Anyway, it's probably not as heavy if you think it further. Also, you should optimize for performance and decrease maintainability only if there is a real reason to do it.

Reply:
Changed this a little bit, so there is no mass update. However, it would be nice to know what is the situation in real life where this is going to be a problem